src/Bicep.LangServer/Handlers/InsertResourceHandler.cs

```diff
@@ -0,0 +1,298 @@
+        }
+
+        private static ResourceDeclarationSyntax CreateResourceSyntax(JsonElement resource, IAzResourceProvider.AzResourceIdentifier resourceId, ResourceTypeReference typeReference)
+        {
+            var properties = new List<ObjectPropertySyntax>();
+            foreach (var property in resource.EnumerateObject())
+            {
+                switch (property.Name.ToLowerInvariant())
+                {
+                    case "id":
+                    case "type":
+                    case "apiVersion":
+                        // Don't add these to the resource properties - they're part of the resource declaration.
+                        break;
+                    case "name":
+                        // Use the fully-qualified name instead of the name returned by the RP.
+                        properties.Add(SyntaxFactory.CreateObjectProperty(
+                            "name",
+                            SyntaxFactory.CreateStringLiteral(resourceId.FullyQualifiedName)));
+                        break;
+                    default:
+                        properties.Add(SyntaxFactory.CreateObjectProperty(
+                            property.Name,
+                            ConvertJsonElement(property.Value)));
+                        break;
+                }
+            }
+
+            var description = SyntaxFactory.CreateDecorator(
+                "description",
+                SyntaxFactory.CreateStringLiteral($"Generated from {resourceId.FullyQualifiedId}"));
+
+            return new ResourceDeclarationSyntax(
+                new SyntaxBase[] { description, SyntaxFactory.NewlineToken, },
+                SyntaxFactory.CreateToken(TokenType.Identifier, "resource"),
+                SyntaxFactory.CreateIdentifier(Regex.Replace(resourceId.UnqualifiedName, "[^a-zA-Z]", "")),
+                SyntaxFactory.CreateStringLiteral(typeReference.FormatName()),
+                null,
+                SyntaxFactory.CreateToken(TokenType.Assignment),
+                SyntaxFactory.CreateObject(properties));
+        }
+
+        private static IAzResourceProvider.AzResourceIdentifier? TryParseResourceId(string? resourceIdString)
+        {
+            if (resourceIdString is null)
+            {
+                return null;
+            }
+
+            if (ResourceId.TryParse(resourceIdString, out var resourceId))
+            {
+                return new(
+                    resourceId.FullyQualifiedId,
+                    resourceId.FormatFullyQualifiedType(),
+                    resourceId.FormatName(),
+                    resourceId.NameHierarchy.Last());
+            }
+
+            var rgRegexOptions = RegexOptions.IgnoreCase | RegexOptions.ExplicitCapture | RegexOptions.Compiled | RegexOptions.CultureInvariant;
+            var rgRegex = new Regex(@"^/subscriptions/(?<subId>[^/]+)/resourceGroups/(?<rgName>[^/]+)$", rgRegexOptions);
```

Comment:
>rgRegex [](http://example.com/codeflow?start=16&length=7)

Btw, do we have the template Url class available to us here?

Reply:
Yeah - TemplateUri is made available by the Deployments codebase. I'm personally not a fan though - regex feels more appropriate.